File: src/main/java/com/grizzly/baseViews/AbstractRelativeLayout.java
package com.grizzly.baseViews;

import android.annotation.TargetApi;
import android.content.Context;
import android.os.Build;
import android.util.AttributeSet;

/**
 * Created by FcoPardo on 3/6/16.
 */
public abstract class AbstractRelativeLayout<T> extends BaseRelativeLayout implements BaseView.OnDataDrivenView<T>{

    protected T data;
    protected Class<T> dataClass;

    public AbstractRelativeLayout(Context context) {
        super(context);
    }

    public AbstractRelativeLayout(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public AbstractRelativeLayout(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    public AbstractRelativeLayout(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
    }

    public <X extends BaseView.OnDataDrivenView<T>> X setData(T aData) {
        this.data = data;
        if(data != null)setControls();
        return (X) this;
    }

    public T getData() {
        return data;
    }

    protected abstract void setControls();
}

File: src/main/kotlin/com/grizzly/baseViews/AbstractRelativeLayout.kt
package com.grizzly.baseViews

import android.content.Context
import android.util.AttributeSet

abstract class AbstractRelativeLayout<T> : BaseRelativeLayout, BaseView.OnDataDrivenView<T> {

    constructor(context : Context) : super(context)
    constructor(context : Context, attrs : AttributeSet) : super(context, attrs)
    constructor(context : Context, attrs : AttributeSet, styleAttr : Int) : super(context, attrs, styleAttr)

    @JvmField
    protected var myData: T? = null
    @JvmField
    protected var dataClass: Class<T>? = null

    override fun setData(aData: T) {
        this.myData = aData
        if (myData != null) setControls()
    }

    override fun getData(): T? {
        return myData
    }

    protected abstract fun setControls()
}